components/prism-core.js

```diff
@@ -140,6 +140,67 @@ var _ = self.Prism = {
 		}
 	},
 
+	optimizeGrammar: function(grammar) {
+		if (!grammar || grammar._competition) {
+			return;
+		}
+
+		var competition = [];
+
+		for (var token in grammar) {
+			if (!grammar.hasOwnProperty(token) || !grammar[token]) {
+				continue;
+			}
+
+			var patterns = grammar[token],
+			    strings = [], rest = [],
+			    regex = {mi: [], i: [], m: [], '': []};
+
+			if (_.util.type(patterns) !== 'Array' || !patterns.length) {
+				if (patterns.competition) {
+					patterns._token = token;
+					competition.push(patterns);
+				}
+				continue;
+			}
+
+			for (var i = 0; i < patterns.length; ++i) {
+				var pattern = patterns[i],
+					type = _.util.type(pattern);
+
+				if (patterns.competition) {
```

Comment:
Shouldn't it be `pattern` (singular) ? We won't set a `competition` property on the array.